### PR TITLE
nlp_seq to linen port

### PR DESCRIPTION
**Pull request by bohnetbd**
I carried out the changes to port the example to linen.

### CODE REVIEW

File: examples/nlp_seq/train.py [hunk 1]
```diff
@@ -17,6 +17,8 @@
 This script trains a Transformer on the Universal dependency dataset.
 """
 
+import tensorflow as tf
```
Comment (andsteing) [inline examples/nlp_seq/train.py:20]: https://www.python.org/dev/peps/pep-0008/#imports

```
Imports should be grouped in the following order:

1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.
You should put a blank line between each group of imports.
```

File: examples/nlp_seq/models.py [hunk 2]
```diff
@@ -14,55 +14,34 @@
...
-    pad_widths = [(0, 0)] * len(x.shape)
-    pad_widths[1] = (1, 0)  # Padding on axis=1
-    padded = jnp.pad(
-        x, pad_widths, mode='constant', constant_values=x.dtype.type(0))
-    return padded[:, :-1]
-  else:
-    # Do nothing in predict mode, as then the sequence length is 1.
-    return x
-
+from flax import linen as nn
```
Comment (andsteing) [inline examples/nlp_seq/models.py:17]: https://www.python.org/dev/peps/pep-0008/#imports

```
Imports should be grouped in the following order:

1. Standard library imports.
2. Related third party imports.
3. Local application/library specific imports.
You should put a blank line between each group of imports.
```

File: examples/nlp_seq/models.py [hunk 3]
```diff
@@ -90,146 +69,150 @@ def init(key, shape, dtype=np.float32):
...
-        x, actual_out_dim, kernel_init=kernel_init, bias_init=bias_init)
-    output = nn.dropout(output, rate=dropout_rate, deterministic=deterministic)
+    cfg = self.config
+    actual_out_dim = (inputs.shape[-1] if self.out_dim is None
+                      else self.out_dim)
+    x = nn.Dense(cfg.mlp_dim,
+                 dtype=cfg.dtype,
+                 kernel_init=cfg.kernel_init,
+                 bias_init=cfg.bias_init)(inputs)
+    x = nn.relu(x)
```
Comment (andsteing) [inline examples/nlp_seq/models.py:141]: `gelu -> relu` : is this intentional ?

File: examples/nlp_seq/models.py [hunk 4]
```diff
@@ -90,146 +69,150 @@ def init(key, shape, dtype=np.float32):
...
-            posemb_init=nn.initializers.normal(stddev=1.0)):
+  Args:
+    config: TransformerConfig dataclass containing hyperparameters.
+  """
+  config: TransformerConfig
+
+  @nn.compact
+  def __call__(self,
+               inputs,
+               inputs_positions=None):
```
Comment (andsteing) [inline examples/nlp_seq/models.py:83]: (Here and elsewhere)

Can we make all arguments to `__call__()` kw-only?
Ideally, we would also add a type annotation.

See also https://github.com/google/flax/issues/231#issuecomment-716568710

File: examples/nlp_seq/models.py [hunk 5]
```diff
@@ -90,146 +69,150 @@ def init(key, shape, dtype=np.float32):
...
-            emb_dim=512,
-            num_heads=8,
-            num_layers=6,
-            qkv_dim=512,
-            mlp_dim=2048,
-            max_len=2048,
-            train=True,
-            dropout_rate=0.3,
-            attention_dropout_rate=0.3):
+            train=False,
```
Comment (andsteing) [inline examples/nlp_seq/models.py:209]: Can we make the additional arguments kw-only and remove the default from the `train` argument?

See also https://github.com/google/flax/issues/231#issuecomment-716568710

File: examples/nlp_seq/models.py [hunk 6]
```diff
@@ -90,146 +69,150 @@ def init(key, shape, dtype=np.float32):
...
+            dropout_rate=0.3):
```
Comment (andsteing) [inline examples/nlp_seq/models.py:210]: Why is this not a setting in `TransformerConfig` ?

File: examples/nlp_seq/models.py [hunk 7]
```diff
@@ -90,146 +69,150 @@ def init(key, shape, dtype=np.float32):
...
-    """Applies Transformer1DBlock module.
+  Args:
+    config: TransformerConfig dataclass containing hyperparameters.
+  """
+  config: TransformerConfig
+
+  @nn.compact
+  def __call__(self,
+               inputs,
+               encoder_mask=None,
```
Comment (andsteing) [inline examples/nlp_seq/models.py:163]: can we make all arguments to `__call__()` kw-only?

See also https://github.com/google/flax/issues/231#issuecomment-716568710

**Review by andsteing — CHANGES_REQUESTED** [review]
Thanks - that will make for the largest tickbox in #568 :-)

Added some initial comments.

File: examples/nlp_seq/train.py (hunk 1 above)
Comment (bohnetbd) [reply examples/nlp_seq/train.py:20]: Thanks. Changed.

File: examples/nlp_seq/models.py (hunk 2 above)
Comment (bohnetbd) [reply examples/nlp_seq/models.py:17]: Thanks! Changed tool

File: examples/nlp_seq/models.py (hunk 3 above)
Comment (bohnetbd) [reply examples/nlp_seq/models.py:141]: no not intentional. changed it back.

File: examples/nlp_seq/models.py (hunk 6 above)
Comment (bohnetbd) [reply examples/nlp_seq/models.py:210]: now taken from TransformerConfig. Thanks!

File: examples/nlp_seq/models.py (hunk 7 above)
Comment (bohnetbd) [reply examples/nlp_seq/models.py:163]: done.

File: examples/nlp_seq/models.py (hunk 6 above)
Comment (bohnetbd) [reply examples/nlp_seq/models.py:210]: use dropout from config.

File: examples/nlp_seq/models.py (hunk 5 above)
Comment (bohnetbd) [reply examples/nlp_seq/models.py:209]: done.

File: examples/nlp_seq/models.py (hunk 4 above)
Comment (bohnetbd) [reply examples/nlp_seq/models.py:83]: done. left type annotations out (for now?)

File: examples/nlp_seq/train.py [hunk 8]
```diff
@@ -405,9 +379,6 @@ def main(argv):
           lambda x: x / eval_denominator,  # pylint: disable=cell-var-from-loop
           eval_metrics_sums)
 
-      # Calculate (clipped) perplexity after averaging log-perplexities:
```
Comment (andsteing) [inline examples/nlp_seq/train.py:408]: why is this removed?
